linera-chain/src/unit_tests/inbox_tests.rs

```diff
@@ -118,14 +130,181 @@ async fn test_inbox_remove_then_add() {
+        .await
+        .unwrap();
+    view.remove_event(&make_unskippable_event(hash, 1, 0, [2]))
+        .await
+        .unwrap();
+    // Inbox is empty again.
+    assert_eq!(view.added_events.count(), 0);
+    assert_eq!(view.removed_events.count(), 0);
+}
+
+#[tokio::test]
+async fn test_inbox_remove_then_add_unskippable() {
+    let hash = CryptoHash::new(&Dummy);
+    let mut view = InboxStateView::new().await;
+    // Remove one event by anticipation.
+    view.remove_event(&make_unskippable_event(hash, 0, 0, [0]))
+        .await
+        .unwrap();
+    // Add the same event
+    view.add_event(make_unskippable_event(hash, 0, 0, [0]))
+        .await
+        .unwrap();
+    // Fail to remove an old event.
+    assert!(matches!(
+        view.remove_event(&make_unskippable_event(hash, 0, 0, [0]))
+            .await,
+        Err(InboxError::IncorrectOrder { .. })
+    ));
+    // Fail to add an old event.
+    assert!(matches!(
+        view.add_event(make_unskippable_event(hash, 0, 0, [0]))
+            .await,
+        Err(InboxError::IncorrectOrder { .. })
+    ));
+    // Remove two more events.
+    view.remove_event(&make_unskippable_event(hash, 0, 1, [1]))
+        .await
+        .unwrap();
+    view.remove_event(&make_unskippable_event(hash, 1, 1, [3]))
+        .await
+        .unwrap();
+    // Fail to add non-matching event.
+    assert!(matches!(
+        view.add_event(make_unskippable_event(hash, 0, 1, [0]))
+            .await,
+        Err(InboxError::UnexpectedEvent { .. })
+    ));
+    // Fail to add non-matching event (hash).
+    assert!(matches!(
+        view.add_event(make_unskippable_event(CryptoHash::new(&Dummy2), 0, 1, [1]))
+            .await,
+        Err(InboxError::UnexpectedEvent { .. })
+    ));
+    // NOT OK to forget about previous consumed events while backfilling.
+    assert!(matches!(
+        view.add_event(make_unskippable_event(hash, 1, 1, [3]))
+            .await,
+        Err(InboxError::UnexpectedEvent { .. })
+    ));
+    // OK to add the two events.
+    view.add_event(make_unskippable_event(hash, 0, 1, [1]))
+        .await
+        .unwrap();
+    // Cannot add an unskippable event that was visibly skipped already.
+    assert!(matches!(
+        view.add_event(make_unskippable_event(hash, 1, 0, [2]))
+            .await,
+        Err(InboxError::UnexpectedEvent { .. })
+    ));
+    view.add_event(make_unskippable_event(hash, 1, 1, [3]))
+        .await
+        .unwrap();
+    // Inbox is empty again.
+    assert_eq!(view.added_events.count(), 0);
+    assert_eq!(view.removed_events.count(), 0);
+}
+
+#[tokio::test]
+async fn test_inbox_add_then_remove_mixed() {
+    let hash = CryptoHash::new(&Dummy);
+    let mut view = InboxStateView::new().await;
+    // Add two events.
+    view.add_event(make_unskippable_event(hash, 0, 1, [1]))
+        .await
+        .unwrap();
+    view.add_event(make_event(hash, 1, 0, [2])).await.unwrap();
+    // Fail to remove non-matching event.
```

Comment:
```suggestion
    // Fail to remove non-matching event (skippability).
```

(At least I had to look twice to realize why it doesn't match.)